### PR TITLE
FORTECH-29: Added silicon skin styles for ToggleButton

### PR DESCRIPTION
**Pull request by daniel-stoian-lgp**
**Feature Added**
Added silicon skin styles for ToggleButton

Enact-DCO-1.0-Signed-off-by: Daniel Stoian daniel.stoian@lgepartner.com

### CODE REVIEW

**Comment by ybsung** [comment]
image

Hello @daniel-stoian-lgp 
Thanks for your contribution.

I found one issue. If adding a long text, the text overflows in the ToggleButton. With the `silicon` skin based on `develop` branch or with other skins, the ToggleButton width extends according to the length of the text.

For example, please check the ToggleButton with the Gallium skin in your PR.
image

Many CSS styles might be changed to fix this issue. So I'm going to review your PR again when you have fixed it.

**Comment by ybsung** [comment]
image

The icon position is strange to me. It is located upper.

**Comment by ybsung** [comment]
In develop branch
image

In the PR branch
image

@daniel-stoian-lgp 
The Button border color when being selected changes. Did you intend to do it?

File: Button/Button.module.less [hunk 1]
```diff
@@ -306,7 +307,8 @@
 			.bg {
 				background-color: @agate-button-active-bg-color;
 				background-image: @agate-button-active-bg-image;
-				border: @agate-button-border-width solid @agate-button-border-color;;
+				border: @agate-button-border-width solid @agate-button-border-color;
+				opacity: @agate-button-active-opacity;
```
Comment (ybsung) [inline Button/Button.module.less:311]: This code breaks the previous button look. Did you intend to do it.

File: Button/Button.module.less [hunk 2]
```diff
@@ -288,6 +288,7 @@
 				background-color: @agate-button-focus-bg-color;
 				background-image: @agate-button-focus-bg-image;
 				border-color: @agate-button-focus-border-color;
+				background-origin: @agate-button-focus-background-origin;
```
Comment (ybsung) [inline Button/Button.module.less:291]: This code breaks the previous button look. Did you intend to do it.

File: ToggleButton/ToggleButton.module.less [hunk 3]
```diff
@@ -61,7 +61,11 @@
 }
 
 .button {
+	display: inline-flex;
```
Comment (ybsung) [inline ToggleButton/ToggleButton.module.less:64]: Why did you add this line instead of `display: inline-block;`? As a component, it seems to be more natural to provide it with the `inline-block` for a button, a checkbox, a text and so on and the `block` for any container like a panel, popup, and so on.

File: ToggleButton/ToggleButton.module.less [hunk 4]
```diff
@@ -70,11 +74,30 @@
...
+			font-size: @agate-togglebutton-indicator-font-size;
+			margin: @agate-togglebutton-indicator-margin;
+		}
+
+		&.minWidth {
+			&,
+			&.huge,
+			&.small,
+			&.smallest {
+				padding: 0 @agate-togglebutton-h-padding;
```
Comment (ybsung) [inline ToggleButton/ToggleButton.module.less:88]: This will affect the ToggleButton in other skins.

File: styles/colors-silicon-day.less [hunk 5]
```diff
@@ -298,17 +302,19 @@
...
-@agate-togglebutton-indicator-off-bg-color: black;
-@agate-togglebutton-indicator-on-bg-color: @agate-highlight;
-@agate-togglebutton-indicator-off-border-color: @agate-button-border-color;
-@agate-togglebutton-indicator-on-border-color: @agate-accent;
+@agate-togglebutton-indicator-off-color: @silicon-dark-gray;
+@agate-togglebutton-indicator-on-color: @silicon-brand;
+@agate-togglebutton-indicator-off-bg-color: transparent;
+@agate-togglebutton-indicator-on-bg-color: transparent;
+@agate-togglebutton-indicator-off-border-color: transparent;
+@agate-togglebutton-indicator-on-border-color: transparent;
```
Comment (ybsung) [inline styles/colors-silicon-day.less:310]: Why do we need to define @agate-togglebutton-indicator-off-border-color and @agate-togglebutton-indicator-on-border-color in styles/colors-silicon-day.less because we already define them in styles/colors-base.less with the same values?

File: styles/colors-silicon-night.less [hunk 6]
```diff
@@ -259,17 +260,19 @@
...
+@agate-togglebutton-indicator-on-bg-color: transparent;
+@agate-togglebutton-indicator-off-border-color: transparent;
+@agate-togglebutton-indicator-on-border-color: transparent;
 @agate-togglebutton-indicator-off-focus-bg-color: @agate-togglebutton-indicator-off-bg-color;
 @agate-togglebutton-indicator-on-focus-bg-color: @agate-togglebutton-indicator-on-bg-color;
 @agate-togglebutton-indicator-off-focus-border-color: @agate-togglebutton-indicator-on-focus-border-color;
-@agate-togglebutton-indicator-on-focus-border-color: @agate-white;
-@agate-togglebutton-indicator-disabled-off-bg-color: @agate-darker-gray;
+@agate-togglebutton-indicator-on-focus-border-color: transparent;
+@agate-togglebutton-indicator-disabled-off-bg-color: transparent;
```
Comment (ybsung) [inline styles/colors-silicon-night.less:273]: Same above.

File: styles/colors-silicon-night.less [hunk 7]
```diff
@@ -259,17 +260,19 @@
...
 @agate-togglebutton-indicator-disabled-on-bg-color: @agate-togglebutton-indicator-disabled-off-bg-color;
-@agate-togglebutton-indicator-disabled-off-border-color: #6c6c6c;
+@agate-togglebutton-indicator-disabled-off-border-color: transparent;
```
Comment (ybsung) [inline styles/colors-silicon-night.less:275]: Same above.

**Comment by daniel-stoian-lgp** [comment]
> In develop branch
> image
> 
> In the PR branch
> image
> 
> @daniel-stoian-lgp
> The Button border color when being selected changes. Did you intend to do it?
> 
> Steps
> 1> Launch Agate Button sampler
> 2> Silicon skin
> 3> Check `selected`

Yes, this is intended. The zeplin specification in Digital cockpit specifies that for Button-day https://zpl.io/aBWj0Zk
The new specification has only night-mode yet, so the day mode styling is left as it was.

**Comment by ybsung** [comment]
@daniel-stoian-lgp 
{quote}
Yes, this is intended. The zeplin specification in Digital cockpit specifies that for Button-day https://zpl.io/aBWj0Zk
The new specification has only night-mode yet, so the day mode styling is left as it was.
{quote}
I got it. Please add that kind of information in GH description or Harmony issue description so that a reviewer and a merger could notice it.

File: Button/Button.module.less (hunk 1 above)
Comment (daniel-stoian-lgp) [reply Button/Button.module.less:311]: yes, the new zeplin spcification for button adds a new opacity for active state for Button. Since ToggleButton is based on Button, I added the fix here
image

File: ToggleButton/ToggleButton.module.less (hunk 4 above)
Comment (daniel-stoian-lgp) [reply ToggleButton/ToggleButton.module.less:88]: I will remove this

File: ToggleButton/ToggleButton.module.less (hunk 3 above)
Comment (daniel-stoian-lgp) [reply ToggleButton/ToggleButton.module.less:64]: I modified the design. See next commit

File: ToggleButton/ToggleButton.module.less [hunk 8]
```diff
@@ -62,22 +62,81 @@
 
 .button {
 	.applySkins({
+		min-width: @agate-togglebutton-min-width;
```
Comment (ybsung) [inline ToggleButton/ToggleButton.module.less:65]: It looks fine to me because it overrides `min-width: @agate-button-height` in Button.module.less and `@agate-button-height` is equal to `@agate-togglebutton-min-width` in all skins except `silicon`.
I just wrote the comment because we could be curious about this code later.

File: ToggleButton/ToggleButton.js [hunk 9]
```diff
@@ -160,6 +169,7 @@ const ToggleButtonBase = kind({
 				{...props}
 				aria-pressed={props.selected}
 				css={css}
+				decoration={props.skin === 'silicon' ? <Icon size={props.size} className={css.toggleIndicator}>circle</Icon> : null}
```
Comment (ybsung) [inline ToggleButton/ToggleButton.js:172]: Hello @daniel-stoian-lgp 

I'm not sure that we should support the indicator with the `decoration` prop.

We support it in other skins except the `silicon` skin with `after` pseudo-element. If it is not our duty to use the `circle` icon, we could implement that part with css styles. Here is my test code to do it ( Only for silicon skin ) - https://github.com/enactjs/agate/commit/a66dd2c89a52846b940e2de413e5c5e32cdb006c.

Please let me know your thought.

CC: @Djspaceg , @seunghoh

Comment (daniel-stoian-lgp) [reply ToggleButton/ToggleButton.js:172]: Hello @ybsung 
Your proposal looks good to me. I don't think also that we need to use a circle icon, but just a circle shaped indicator
Maybe the others have some comments. Otherwise, i will apply your proposal.

Comment (ybsung) [reply ToggleButton/ToggleButton.js:172]: I tried to reduce the css with &.huge {}, &.small {}, and &.smallest {}. But for now there is no idea. Let's revisit it after fixing above.

File: Button/Button.module.less (hunk 2 above)
Comment (Djspaceg) [reply Button/Button.module.less:291]: I think maybe this should just be simplified to `background-origin: @agate-button-background-origin;` and assigned to the regular resting button, rather than specifically on the `focus` state. It's not previously set, so having it only apply on focus seems incomplete.

File: styles/variables-silicon.less [hunk 10]
```diff
@@ -216,6 +216,20 @@
...
+@agate-togglebutton-small-client-padding-left: @agate-togglebutton-client-padding-left;
+@agate-togglebutton-smallest-min-width: 90px;
+@agate-togglebutton-smallest-client-padding-left: @agate-togglebutton-client-padding-left;
+@agate-togglebutton-indicator-border-radius: 50%;
+@agate-togglebutton-indicator-size: 30px;
+@agate-togglebutton-indicator-position: 50% auto auto 30px;
+@agate-togglebutton-indicator-height: 15px;
+@agate-togglebutton-indicator-transform: translateY(-50%);
+@agate-togglebutton-indicator-width: @agate-togglebutton-indicator-height;
+@agate-togglebutton-huge-indicator-position: 50% auto auto 60px;
```
Comment (Djspaceg) [inline styles/variables-silicon.less:233]: If the transform included X, you could simplify the positioning a lot.
```css
@agate-togglebutton-indicator-transform: translate(-50%, -50%);
...
@agate-togglebutton-indicator-position: 50% auto auto 50%
@agate-togglebutton-huge-indicator-position: @agate-togglebutton-indicator-position;
@agate-togglebutton-small-indicator-position: @agate-togglebutton-indicator-position;
@agate-togglebutton-smallest-indicator-position: @agate-togglebutton-indicator-position;
```
Then you wouldn't even need the `huge`, `small`, and `smallest` position variables or their accompanying CSS either, since they'd just all be the same.

Comment (daniel-stoian-lgp) [reply styles/variables-silicon.less:233]: thanks for the tip.

Comment (daniel-stoian-lgp) [reply styles/variables-silicon.less]: i solved the issue by leaving auto value for position-left
@agate-togglebutton-indicator-position: 50% auto auto;

**Comment by ybsung** [comment]
### Checklist

- [x] npm run parse - Pass ( Known parse errors: ThumbnailItem, useChainRefs )
- [x] npm run test-ss - Pass ( for more information about how to run, please check my comment - https://github.com/enactjs/agate/pull/274#issuecomment-646705673 )

File: styles/colors-silicon-day.less [hunk 11]
```diff
@@ -300,17 +301,17 @@
 
 // Toggle Colors
 // ---------------------------------------
-@agate-togglebutton-indicator-off-bg-color: black;
-@agate-togglebutton-indicator-on-bg-color: @agate-highlight;
-@agate-togglebutton-indicator-off-border-color: @agate-button-border-color;
-@agate-togglebutton-indicator-on-border-color: @agate-accent;
+@agate-togglebutton-indicator-off-bg-color: @silicon-dark-gray;
+@agate-togglebutton-indicator-on-bg-color: @silicon-accent;
```
Comment (Djspaceg) [inline styles/colors-silicon-day.less:305]: I think this one perhaps should have been `@agate-accent`.

It may not be good to have variables so close in name with very different functionality.

Comment (daniel-stoian-lgp) [reply styles/colors-silicon-day.less]: i don't know how I missed that one. changed it to agate-accent
